graphql_compiler/schema/schema_info.py

```diff
@@ -1,13 +1,116 @@
+from sqlalchemy.engine.interfaces import Dialect
 
 from . import is_vertex_field_name
 
 
+# Describes the intent to join two tables using the specified columns.
+#
+# The resulting join expression could be something like:
+# JOIN origin_table.from_column = destination_table.to_column
+#
+# The type of join (inner vs left, etc.) is not specified.
+# The tables are not specified.
+DirectJoinDescriptor = namedtuple(
+    "DirectJoinDescriptor",
+    (
+        "from_column",  # The column in the source table we intend to join on.
+        "to_column",  # The column in the destination table we intend to join on.
+    ),
+)
+
+
+@dataclass
+class SchemaInfo:
+    """Class to store all schema info required for querying."""
+
+    # This is a required component of any SchemaInfo objects.
+    schema: GraphQLSchema
+
+    # Optional dict of GraphQL interface or type -> GraphQL union.
+    # Used as a workaround for GraphQL's lack of support for
+    # inheritance across "types" (i.e. non-interfaces), as well as a
+    # workaround for Gremlin's total lack of inheritance-awareness.
+    # The key-value pairs in the dict specify that the "key" type
+    # is equivalent to the "value" type, i.e. that the GraphQL type or
+    # interface in the key is the most-derived common supertype
+    # of every GraphQL type in the "value" GraphQL union.
+    # Recursive expansion of type equivalence hints is not performed,
+    # and only type-level correctness of this argument is enforced.
+    # See README.md for more details on everything this parameter does.
+    # *****
+    # Be very careful with this option, as bad input here will
+    # lead to incorrect output queries being generated.
+    # *****
+    type_equivalence_hints: Dict[str, str]  # Optional
+
+    # ####
+    # The fields below are required for SQL databases.
+    # ####
+    # Specifying the dialect for which we are compiling, if a dialect exists
+    # e.g. sqlalchemy.dialects.mssql.dialect()
+    dialect: Dialect
+    # dict mapping every GraphQL object type or interface type name in the schema to
+    # a sqlalchemy table.
+    # Column types that do not exist for this dialect are not allowed.
+    # All tables are expected to have primary keys.
+    vertex_name_to_table: Dict[str, sqlalchemy.Table]
+    # dict mapping every GraphQL object type or interface type name in the schema to
+    # dict mapping every vertex field name at that type to a DirectJoinDescriptor.
+    # The tables the join is to be performed on are not specified.
+    # They are inferred from the schema and the tables dictionary.
+    join_descriptors: Dict[str, Dict[str, DirectJoinDescriptor]]
+
+    # TODO: consider extending this to being able to support QueryPlanningSchemaInfo
+
+    # TODO: consider - maybe move this outside of the class for easy import for users?
+    # e.g. it would be nice if folks could do graphql_compiler.match import create_match_schema_info
```

Comment:
Yes, I think they should be outside of here and part of the respective backend.

One nit: while the backend is `MATCH` to differentiate it from `Gremlin`, it would probably be a good idea to make the import be `graphql_compiler.orientdb` based on the vendor to make it easier to find.